### ISSUE TITLE
Missing / Too much string escaping?

### ISSUE THREAD

**Issue by itsho** [Tyrrrz/CliWrap#72]
Hello.

The following code (contains RANDOM string):
```
var cmdTest = Cli.Wrap("myexe")
    .WithArguments(a => a
        .Add(new[]
            {
                "/superstring",
                "ogq77a3auubgQaVHUXhKpVa*NR!YTsEjn8MA2^9$gZwb2D2z3EU^Fk7d8pS&psPg9F8M*jbfyCDWsewF$5osz9LmcsGUf5jV^@WarHuYPxZXN@GU5&AyhE!3t%W&f@ra"
                }));
```

produces this command argument - **no quotes around the superstring** while I expect it to be escaped:
```
/superstring ogq77a3auubgQaVHUXhKpVa*NR!YTsEjn8MA2^9$gZwb2D2z3EU^Fk7d8pS&psPg9F8M*jbfyCDWsewF$5osz9LmcsGUf5jV^@WarHuYPxZXN@GU5&AyhE!3t%W&f@ra
```

But, if i want to add a single quotation mark with this code:
```
var cmdTest = Cli.Wrap("myexe")
	.WithArguments(a => a
		.Add(new[]
		{
			"/superstring",
			"\"ogq77a3auubgQaVHUXhKpVa*NR!YTsEjn8MA2^9$gZwb2D2z3EU^Fk7d8pS&psPg9F8M*jbfyCDWsewF$5osz9LmcsGUf5jV^@WarHuYPxZXN@GU5&AyhE!3t%W&f@ra\""
		}));
```

I get the following command arguments - **too many quotes**:
```
/superstring "\"ogq77a3auubgQaVHUXhKpVa*NR!YTsEjn8MA2^9$gZwb2D2z3EU^Fk7d8pS&psPg9F8M*jbfyCDWsewF$5osz9LmcsGUf5jV^@WarHuYPxZXN@GU5&AyhE!3t%W&f@ra\""
```

My way to overcome this is to create the arguments by myself with `StringBuilder` and send it to the `WithArguments` method.

Is there another way to solve it?

Thanks!

**Comment by Tyrrrz** [comment]
The quotes need to be escaped because they have a special meaning in command line applications.

**Comment by itsho** [comment]
Hi.
I'm well aware of that. Please let me rephrase my question:

How can I add only one quotation mark on each side of my string?

**Comment by Tyrrrz** [comment]
Why do you want to add quotation marks? If they are necessary, they will be added automatically.
If you wish to specify the command line directly, then you can format it as a string yourself and then use `WithArguments(string)` overload.

**Comment by itsho** [comment]
They are nessacary and was not added as you can see in my first example.

The string contains many chars that requires escaping and yet, escaping was not done.

**Comment by Tyrrrz** [comment]
Which characters do you think should have been escaped?

**Comment by itsho** [comment]
Both: "&", "^"

also, I guess that some people would also like to escape "%".

![CharsToEscape](https://user-images.githubusercontent.com/1774797/85229153-f8728880-b3f0-11ea-839b-ee9da6b30c81.png)

**Comment by Tyrrrz** [comment]
I believe these are characters with special meanings only in shell, i.e. `cmd` or `bash`. When you pass arguments to an application directly, those are just regular characters like any other. Are you having issues with trying to wrap `cmd`?

**Comment by itsho** [comment]
I'm having a hard time explaining the issue.
hopefully, the following example will demonstrate it better than words:

## C++ app to demonstrate the issue
 * given the following simple C++ console app:
```cpp
#include <<iostream>
using namespace std;
int main(const int argc, char** argv)
{
    std::cout << "CppConsoleApp argument tester" << endl;
    for (size_t i = 0; i < static_cast<size_t>(argc); i++)
    {
        std::cout << "Argument " << i << " :" << argv[i] << endl;
    }
}
```

## Mixed results examples
* When trying to run the compiled app in `cmd`, these are the results:
```
>CppConsoleApp.exe a b c
CppConsoleApp argument tester
Argument 0 :CppConsoleApp.exe
Argument 1 :a
Argument 2 :b
Argument 3 :c

>CppConsoleApp.exe a b c^
More?

>CppConsoleApp.exe a b c^^
CppConsoleApp argument tester
Argument 0 :CppConsoleApp.exe
Argument 1 :a
Argument 2 :b
Argument 3 :c^

>CppConsoleApp.exe a b c&&
The syntax of the command is incorrect.

>CppConsoleApp.exe a b c%%
CppConsoleApp argument tester
Argument 0 :CppConsoleApp.exe
Argument 1 :a
Argument 2 :b
Argument 3 :c%%
```

as you can see - only the first case is valid. 
cases 2-4 (included) are wrong.
case 5 is exceptional since it **seems** to be working, but it WON'T work if we will use a variable name like "%TEMP%"

## Screenshot
![2020-06-23 23_35_34-C__Windows_System32_cmd exe](https://user-images.githubusercontent.com/1774797/85459045-7332d380-b5aa-11ea-8bca-084b7b23426b.png)

**Comment by Tyrrrz** [comment]
Makes sense. Can you also explain what exactly are you doing with CliWrap? Are you wrapping `CppConsoleApp.exe` directly or `cmd` that runs `CppConsoleApp.exe`?

**Comment by itsho** [comment]
I tried both directly and indirectly - got the same result.

**Comment by Tyrrrz** [comment]
It should definitely work directly, but I can imagine issues if you're wrapping it indirectly.
What happens when you wrap it directly and execute with `a b c^`? It shouldn't write `More?` because that comes from the shell itself.

**Comment by itsho** [comment]
When I use the dummy c++ app - you are totally right, there IS a difference between direct and indirect.
Direct - works as expected.
indirect - it fails to pass the whole string.

But, there's a specific software I'm using which does not behave the same.
I suspect that the app is wrapping `cmd` in someway, which forces me to escape the strings I'm sending.

I tried multiple ways, and the following issue is the exact same issue I'm facing while wrapping this specific software:
```csharp
const string COMMAND_NAME = @"cmd";
const string ARGUMENT = @"/c 3rdParty\DummyConsoleApp.exe a ""b^b"" c^c d$ e^ /superstring ogq77a3auubgQaVHUXhKpVa*NR!YTsEjn8MA2^9$gZwb2D2z3EU^Fk7d8pS&psPg9F8M*jbfyCDWsewF$5osz9LmcsGUf5jV^@WarHuYPxZXN@GU5&AyhE!3t%W&f@ra";
var cmd = Cli.Wrap(COMMAND_NAME)
	.WithValidation(CommandResultValidation.None)
	.WithArguments(ARGUMENT);

var res = cmd.ExecuteBufferedAsync();

var stdOutput= res.Task.Result.StandardOutput;
var errOutput = res.Task.Result.StandardError;
```

The StandardOutput is
> DummyConsoleApp argument tester
> Argument 0: a
> Argument 1: b^b
> Argument 2: cc
> Argument 3: d$
> Argument 4: e
> Argument 5: /superstring
> Argument 6: ogq77a3auubgQaVHUXhKpVa*NR!YTsEjn8MA29$gZwb2D2z3EUFk7d8pS

And the StandardError is:
> 'psPg9F8M*jbfyCDWsewF$5osz9LmcsGUf5jV@WarHuYPxZXN@GU5' is not recognized as an internal or external command,
> operable program or batch file.
> 'AyhE!3t%W' is not recognized as an internal or external command,
> operable program or batch file.
> 'f@ra' is not recognized as an internal or external command,
> operable program or batch file.

**Comment by Tyrrrz** [comment]
I see.
Would a bool switch to disable automatic escaping work for you?
Something like: `cmd.WithArguments(new[] {"foo", "bar"}, false)`?

**Comment by itsho** [comment]
Yes. 
And I think that the second one is better since it gives more control.

Also, if possible, can you add an overload with Func<string,string> to let the uses utilize his own escaping method?

thank you very much.
